In [1]:
"""
A few decisions.

###########################################################
# Parsing the answer:
###########################################################
We take the answer by taking the last integer in the answer string.
If we don't find an integer, we call word2num on the answer string, to maybe get a word form number.
If we don't find a number, in the majority vote case, we ignore that generation in the vote.
If none of the generations have a number, we return the default value.
Similarily, in the case without majority vote, if we don't find a number, we return the default value.

It would be fair to say that we should maybe not use the default value, and just assign the answer as incorrect.
We may do that eventually. It would be pretty straightfoward, just change the default value to a stopper object,
the comparison with the reference answer would always be false.

Floating point numbers:
We currently round to the closest integer.


###########################################################
# Picking questions in the dataset:
###########################################################
More than 85% of the questions in the ASDiv dataset have a single, integer answer.
We only consider those questions, as they make the task of parsing the answer easier.

"""


import collections

import importlib
import itertools
import math
import more_itertools
from pathlib import Path
import random
import re
import time
import xml

import datasets
import matplotlib.pyplot as plt
import numpy as np
import rich
import rich.table
import torch
from tqdm import tqdm
import transformers
import wget
from text2digits import text2digits

import general_utils as utils

import asdiv_dataset

importlib.reload(text2digits)

<module 'text2digits.text2digits' from '/home/mila/g/gagnonju/.main_anaconda/lib/python3.9/site-packages/text2digits/text2digits.py'>

In [2]:
# dataset_train = datasets.load_dataset("gsm8k", "main", split="train")
# dataset_test  = datasets.load_dataset("gsm8k", "main", split="test")

# def clean_text(sample):
#     return {k: v.replace("<<", "(").replace(">>", ")").strip() for k, v in sample.items()}

# def split_answer_scratchpad(sample):
#     scratchpad, answer = sample["answer"].split("####")
#     return {
#         "question": sample["question"].strip(), 
#         "answer": answer.strip(), 
#         "scratchpad": scratchpad.strip()
#     }


# dataset_train = dataset_train.map(clean_text).map(split_answer_scratchpad)
# dataset_test  = dataset_test .map(clean_text).map(split_answer_scratchpad)

# print(dataset_train[0].keys())
# print(dataset_test[0].keys())


In [3]:
dataset_train = asdiv_dataset.ASDivInteger(cache_path="ASDiv.xml", quiet=False)
dataset_test  = dataset_train

Parsing dataset...
Parsing complete.
kept 87.8% of data points


In [4]:
###############################################################################
# Load the model
###############################################################################
model_name = "google/flan-t5-xxl"
tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)

with utils.ctx_timeit(f"Loading model `{model_name}`"):
    model = transformers.AutoModelForSeq2SeqLM.from_pretrained(model_name)

with utils.ctx_timeit("Converting model's type"):
    # model_cpu
    pass

with utils.ctx_timeit(f"Moving model to GPU"):
    model = model.cuda()

rich.print(f"[bold blue]Model dtype:[/]  {model.dtype}")
devices = collections.Counter(x.device.type for x in model.parameters())
rich.print(f"\n[bold blue]Model device:[/] {devices}")

assert len(devices) == 1 and "cuda" in devices, devices

Loading model `google/flan-t5-xxl` took 115.74717 seconds

Converting model's type took 0.00000 seconds

Moving model to GPU took 10.67501 seconds

Model dtype:  torch.float32

Model device: Counter({'cuda': 558})

In [5]:
def compose_fewshot_context(dataset, n, with_scratchpad, seed):
    """ Creates a random few-shot context. Works fine with n = 0. """
    rng = random.Random(seed)

    indices = rng.sample(range(len(dataset)), n)
    output = []
    for i in indices:
        scratchpad = dataset[i]["scratchpad"]
        answer = dataset[i]["answer"]

        assert "#" not in scratchpad, scratchpad
        assert "#" not in answer, answer

        output.append(
            "Question:\n" + dataset[i]["question"] + 
            ("\nScratchpad:\n" + scratchpad if with_scratchpad else "") + 
            "\nAnswer:\n" + answer
        )

    return "\n\n".join(output)


def collate(inputs, tokenizer, few_shot_context, with_scratchpad):
    """ Collates the inputs and prompts into a single list of strings. """
    # first_context_addition = few_shot_context + "\n\nQuestion:\n"
    # final_context_addition = "\nScratchpad:\n" if with_scratchpad else "\nAnswer:\n"
    first_context_addition = ""
    final_context_addition = " Let's think about it step by step."

    inputs = utils.dict_unzip(inputs)

    output = tokenizer(
        [
            first_context_addition + question + final_context_addition 
            for question in inputs["question"]
        ],
        padding=True,
        return_tensors="pt"
    ) 

    output["answer"] = inputs["answer"]
    output["scratchpad"] = inputs["scratchpad"]
    
    return {
        k: v.to("cuda") if isinstance(v, torch.Tensor) else v 
        for k, v in output.items()
    }


###############################################################################
# Dataset stuff
###############################################################################
def format_output(output):
    try:
        float_conversion = float(output.replace(",", ""))
    except (ValueError, ArithmeticError) as err:
        rich.print(
            f"[bold red]Failed to convert to float. "
            f"value is:[/] `{output}`, "
            f"[bold red]error is:[/] {type(err)} {err}"
        )
        return "0"
    
    rounding = round(float_conversion)
    return str(rounding)


def majority_vote(generated, tokenizer, answer_extraction_fn, verbose):
    answers = []
    for entry in generated:
        decoded = tokenizer.decode(entry, skip_special_tokens=True)
        output = answer_extraction_fn(decoded)
        if output is not None:
            answers.append(format_output(output))
    
    counter = collections.Counter(answers)
    if verbose:
        print(counter)
    if counter:
        return counter.most_common(1)[0][0]
    else:
        return "0"


def majority_vote_batch(generated, tokenizer, answer_extraction_fn, verbose):
    for entry in generated:
        yield majority_vote(entry, tokenizer, answer_extraction_fn, verbose)

In [6]:

text2digits_ = text2digits.Text2Digits()
num_pat = re.compile(r"\d+(?:[\,\.]\d+)?")


def deal_with_words(text):
    converted = text2digits_.convert(text)
    output = num_pat.findall(converted)[-1]
    rich.print(
        f"[bold blue]text2digits[/]:\n"
        f" \t - [green]source:[/]    {text}\n"
        f" \t - [green]converted:[/] {converted}\n"
        f" \t - [green]final:[/]     {output}"
    )
    return output 


def split_fn(x):
    results = num_pat.findall(x)
    if not results:
        try:
            output = deal_with_words(x)
        except ValueError:
            output = None

        if output is not None:
            rich.print(f"[red]split_fn: no numbers found. Received:[/] `{x}`. Word2Num worked. Output: `{output}`")
            output = str(output)
        else:
            rich.print(f"[red]split_fn: no numbers found. Received:[/] `{x}`")
            output = None
    else:
        output = results[-1]
    return output


def compare(pred, answ):
    return format_output(pred.strip()) == format_output(answ.strip())


def run(
    *,
    shuffle,
    verbose,
    context,
    num_beams,
    batch_size,
    max_new_tokens,
    with_scratchpads,
    use_majority_vote,
    use_group_beam_search,
    generation_extra_kwargs
):
    args = locals().copy()
    
    with torch.inference_mode():
        dataloader = torch.utils.data.DataLoader(
            dataset_test,
            shuffle=shuffle,
            batch_size=batch_size,
            collate_fn=lambda inputs: collate(
                inputs, tokenizer, context, with_scratchpads
            )
        )

        outputs = []
        tqdm_obj = tqdm(dataloader)

        extra_kwargs = generation_extra_kwargs.copy()
        if use_group_beam_search:
            extra_kwargs["num_beam_groups"] = num_beams

        for batch in tqdm_obj:
            output = model.generate(
                input_ids            = batch["input_ids"],
                attention_mask       = batch["attention_mask"],
                num_beams            = num_beams,
                num_return_sequences = num_beams if use_majority_vote else 1,
                max_new_tokens       = max_new_tokens,
                **extra_kwargs
            ).reshape(batch_size, num_beams if use_majority_vote else 1, -1)

            predictions = list(majority_vote_batch(output, tokenizer, split_fn, verbose))
            raw_decoded = [
                [tokenizer.decode(x, skip_special_tokens=True) for x in batch_entry] 
                for batch_entry in output
            ]

            answer_decoded = [list(map(split_fn, x)) for x in raw_decoded]

            if verbose:
                for prediction, answer, raw_decoded_entry, answer_decoded, input_text in zip(
                    predictions, 
                    batch["answer"], 
                    raw_decoded, 
                    answer_decoded, 
                    [tokenizer.decode(x) for x in batch["input_ids"]]
                ):

                    rich.print(
                        f"[bold blue]input_text[/]:      {input_text}\n"
                        f"[bold blue]ref_answer[/]:      {answer}\n"
                        f"[bold blue]prediction[/]:      {prediction}\n"
                        # f"[bold blue]raw_decoded[/]:     {raw_decoded_entry}\n"
                        f"[bold blue]answer_decoded[/]:  {answer_decoded}"
                    )
                    rich.print("[bold blue]Raw Decoded:[/]")
                    for v in raw_decoded_entry:
                        rich.print(f" [bold]-[/] {v}")

            good_bad_preds = [
                compare(pred=pred, answ=answ) 
                for pred, answ in zip(predictions, batch["answer"])
            ]

            # print([(format_output(a), format_output(b), a, b) for a, b in zip(predictions, batch["answer"])])
            # print(good_bad_preds)

            outputs.extend(good_bad_preds)
            tqdm_obj.set_description(f"Accuracy: {np.mean(outputs):.1%}")
            
        accuracy = np.mean(outputs)

        rich.print(args)
        rich.print(f"[bold green]Accuracy, {model.dtype}: {accuracy:.1%}")


verbose = False
shuffle = True
n_shots = 0
num_beams = 8
batch_size = 1
max_new_tokens = 100
with_scratchpads = True
use_majority_vote = True
use_group_beam_search = False
few_shot_context_rng_seed = 42  # Makes sure the context is the same if we want it to stay the same

generation_extra_kwargs = dict(
    repetition_penalty=50.,
)


context = compose_fewshot_context(
        dataset_train, 
        n_shots, 
        with_scratchpads, 
        few_shot_context_rng_seed,
    )


print(context)
assert not context



run(
    shuffle=shuffle,
    verbose=verbose,
    context=context,
    num_beams=num_beams, 
    batch_size=batch_size, 
    max_new_tokens=max_new_tokens, 
    with_scratchpads=with_scratchpads, 
    use_majority_vote=use_majority_vote, 
    use_group_beam_search=use_group_beam_search,
    generation_extra_kwargs=generation_extra_kwargs
)

Accuracy: 33.9%:  15%|█▍        | 301/2023 [09:29<48:45,  1.70s/it]  

text2digits:
         - source:    The sum of three consecutive even numbers is one hundred twenty-three. What is the largest of
the three numbers?
         - converted: The sum of 3 consecutive even numbers is 123. What is the largest of the 3 numbers
         - final:     3

split_fn: no numbers found. Received: `The sum of three consecutive even numbers is one hundred twenty-three. What 
is the largest of the three numbers?`. Word2Num worked. Output: `3`

text2digits:
         - source:    The sum of three consecutive even numbers is one hundred twenty-three. What is the smallest 
of the three numbers?
         - converted: The sum of 3 consecutive even numbers is 123. What is the smallest of the 3 numbers
         - final:     3

split_fn: no numbers found. Received: `The sum of three consecutive even numbers is one hundred twenty-three. What 
is the smallest of the three numbers?`. Word2Num worked. Output: `3`

text2digits:
         - source:    The sum of three consecutive even numbers is one hundred twenty-three. What is the largest of
the three numbers?
         - converted: The sum of 3 consecutive even numbers is 123. What is the largest of the 3 numbers
         - final:     3

split_fn: no numbers found. Received: `The sum of three consecutive even numbers is one hundred twenty-three. What 
is the largest of the three numbers?`. Word2Num worked. Output: `3`

text2digits:
         - source:    The sum of three consecutive even numbers is one hundred twenty-three. What is the smallest 
of the three numbers?
         - converted: The sum of 3 consecutive even numbers is 123. What is the smallest of the 3 numbers
         - final:     3

split_fn: no numbers found. Received: `The sum of three consecutive even numbers is one hundred twenty-three. What 
is the smallest of the three numbers?`. Word2Num worked. Output: `3`

Accuracy: 37.1%:  25%|██▍       | 499/2023 [15:26<36:20,  1.43s/it]  

text2digits:
         - source:    The store has fifty-six + eleven = sixty shirts now. The answer: sixty.
         - converted: The store has 56 + 11 = 60 shirts now. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The store has fifty-six + eleven = sixty shirts now. The answer: sixty.`. 
Word2Num worked. Output: `60`

text2digits:
         - source:    The store has fifty-six + eleven = sixty shirts now. So the answer is sixty.
         - converted: The store has 56 + 11 = 60 shirts now. So the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The store has fifty-six + eleven = sixty shirts now. So the answer is 
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The store has fifty-six + eleven = sixty shirts now. The answer: sixty.
         - converted: The store has 56 + 11 = 60 shirts now. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The store has fifty-six + eleven = sixty shirts now. The answer: sixty.`. 
Word2Num worked. Output: `60`

text2digits:
         - source:    The store has fifty-six + eleven = sixty shirts now. So the answer is sixty.
         - converted: The store has 56 + 11 = 60 shirts now. So the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The store has fifty-six + eleven = sixty shirts now. So the answer is 
sixty.`. Word2Num worked. Output: `60`

Accuracy: 37.6%:  30%|██▉       | 601/2023 [18:42<29:30,  1.24s/it]  

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = eighty three. The answer: 
eighty three.
         - converted: The total number of people in line is 60 - 1 + 22 = 83. The answer: 83
         - final:     83

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = eighty 
three. The answer: eighty three.`. Word2Num worked. Output: `83`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety three. The answer: 
ninety three.
         - converted: The total number of people in line is 60 - 1 + 22 = 93. The answer: 93
         - final:     93

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety 
three. The answer: ninety three.`. Word2Num worked. Output: `93`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = seventy three. The answer: 
seventy three.
         - converted: The total number of people in line is 60 - 1 + 22 = 73. The answer: 73
         - final:     73

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = seventy 
three. The answer: seventy three.`. Word2Num worked. Output: `73`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety. The answer: ninety.
         - converted: The total number of people in line is 60 - 1 + 22 = 90. The answer: 90
         - final:     90

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety. The
answer: ninety.`. Word2Num worked. Output: `90`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = eighty. The answer: eighty.
         - converted: The total number of people in line is 60 - 1 + 22 = 80. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = eighty. The
answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety. So the answer is 
ninety.
         - converted: The total number of people in line is 60 - 1 + 22 = 90. So the answer is 90
         - final:     90

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety. So 
the answer is ninety.`. Word2Num worked. Output: `90`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = eighty three. The answer: 
eighty three.
         - converted: The total number of people in line is 60 - 1 + 22 = 83. The answer: 83
         - final:     83

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = eighty 
three. The answer: eighty three.`. Word2Num worked. Output: `83`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety three. The answer: 
ninety three.
         - converted: The total number of people in line is 60 - 1 + 22 = 93. The answer: 93
         - final:     93

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety 
three. The answer: ninety three.`. Word2Num worked. Output: `93`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = seventy three. The answer: 
seventy three.
         - converted: The total number of people in line is 60 - 1 + 22 = 73. The answer: 73
         - final:     73

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = seventy 
three. The answer: seventy three.`. Word2Num worked. Output: `73`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety. The answer: ninety.
         - converted: The total number of people in line is 60 - 1 + 22 = 90. The answer: 90
         - final:     90

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety. The
answer: ninety.`. Word2Num worked. Output: `90`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = eighty. The answer: eighty.
         - converted: The total number of people in line is 60 - 1 + 22 = 80. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = eighty. The
answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    The total number of people in line is sixty - one + twenty two = ninety. So the answer is 
ninety.
         - converted: The total number of people in line is 60 - 1 + 22 = 90. So the answer is 90
         - final:     90

split_fn: no numbers found. Received: `The total number of people in line is sixty - one + twenty two = ninety. So 
the answer is ninety.`. Word2Num worked. Output: `90`

Accuracy: 38.2%:  30%|███       | 608/2023 [18:50<30:34,  1.30s/it]

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Before Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Before Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Before Making Pie`. 
Word2Num worked. Output: `147`

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left After Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Left After Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left After Making Pie`. 
Word2Num worked. Output: `147`

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left Before Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Left Before Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left Before Making Pie`. 
Word2Num worked. Output: `147`

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Before Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Before Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Before Making Pie`. 
Word2Num worked. Output: `147`

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left After Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Left After Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left After Making Pie`. 
Word2Num worked. Output: `147`

text2digits:
         - source:    The chef has fifty-eight cherries left after making the pie, so he must have had at least 
thirty nine cherries before that. Since there are two ways to find this number we can calculate it as follows: 
Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left Before Making Pie
         - converted: The chef has 58 cherries left after making the pie, so he must have had at least 39 cherries 
before that. Since there are 2 ways to find this number we can calculate it as follows: 39 Cherries + 58 Cherries =
147 Cherry's Left Before Making Pie
         - final:     147

split_fn: no numbers found. Received: `The chef has fifty-eight cherries left after making the pie, so he must have
had at least thirty nine cherries before that. Since there are two ways to find this number we can calculate it as 
follows: Thirty Nine Cherries + Fifty Eight Cherries = One Hundred Forty Seven Cherry's Left Before Making Pie`. 
Word2Num worked. Output: `147`

Accuracy: 38.2%:  30%|███       | 613/2023 [19:00<37:45,  1.61s/it]

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start. The answer: sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start. The answer: 
sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his game. The answer: sixty 
nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his game. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his game. The 
answer: sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his video game. The answer: 
sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his video game. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his video 
game. The answer: sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start. So the answer is sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start. So the answer is 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start. So the answer is
sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his game. So the answer is 
sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his game. So the answer is 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his game. So 
the answer is sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start. The answer: sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start. The answer: 
sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his game. The answer: sixty 
nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his game. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his game. The 
answer: sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his video game. The answer: 
sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his video game. The answer: 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his video 
game. The answer: sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start. So the answer is sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start. So the answer is 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start. So the answer is
sixty nine.`. Word2Num worked. Output: `69`

text2digits:
         - source:    Ned had seventy - thirteen = sixty nine lives at the start of his game. So the answer is 
sixty nine.
         - converted: Ned had 70 - 13 = 69 lives at the start of his game. So the answer is 69
         - final:     69

split_fn: no numbers found. Received: `Ned had seventy - thirteen = sixty nine lives at the start of his game. So 
the answer is sixty nine.`. Word2Num worked. Output: `69`

Accuracy: 38.8%:  39%|███▉      | 787/2023 [24:10<31:02,  1.51s/it]  

text2digits:
         - source:    Isabel gave eighty - seven games to her friend, so she would have ninety - eighty = one game 
left. The answer: one.
         - converted: Isabel gave 80 - 7 games to her friend, so she would have 90 - 80 = 1 game left. The answer: 
1
         - final:     1

split_fn: no numbers found. Received: `Isabel gave eighty - seven games to her friend, so she would have ninety - 
eighty = one game left. The answer: one.`. Word2Num worked. Output: `1`

text2digits:
         - source:    Isabel gave eighty-seven DS games to her friend, so she would have ninety - eighty seven = 
three left. The answer: three.
         - converted: Isabel gave 87 DS games to her friend, so she would have 90 - 87 = 3 left. The answer: 3
         - final:     3

split_fn: no numbers found. Received: `Isabel gave eighty-seven DS games to her friend, so she would have ninety - 
eighty seven = three left. The answer: three.`. Word2Num worked. Output: `3`

text2digits:
         - source:    Isabel gave eighty-seven DS games to her friend, so she would have ninety - eighty seven = 
one game left. The answer: one.
         - converted: Isabel gave 87 DS games to her friend, so she would have 90 - 87 = 1 game left. The answer: 1
         - final:     1

split_fn: no numbers found. Received: `Isabel gave eighty-seven DS games to her friend, so she would have ninety - 
eighty seven = one game left. The answer: one.`. Word2Num worked. Output: `1`

text2digits:
         - source:    Isabel gave eighty-seven games to her friend, so she would have ninety - eighty seven = three
left. The answer: three.
         - converted: Isabel gave 87 games to her friend, so she would have 90 - 87 = 3 left. The answer: 3
         - final:     3

split_fn: no numbers found. Received: `Isabel gave eighty-seven games to her friend, so she would have ninety - 
eighty seven = three left. The answer: three.`. Word2Num worked. Output: `3`

text2digits:
         - source:    Isabel gave eighty - seven games to her friend, so she would have ninety - eighty = one game 
left. The answer: one.
         - converted: Isabel gave 80 - 7 games to her friend, so she would have 90 - 80 = 1 game left. The answer: 
1
         - final:     1

split_fn: no numbers found. Received: `Isabel gave eighty - seven games to her friend, so she would have ninety - 
eighty = one game left. The answer: one.`. Word2Num worked. Output: `1`

text2digits:
         - source:    Isabel gave eighty-seven DS games to her friend, so she would have ninety - eighty seven = 
three left. The answer: three.
         - converted: Isabel gave 87 DS games to her friend, so she would have 90 - 87 = 3 left. The answer: 3
         - final:     3

split_fn: no numbers found. Received: `Isabel gave eighty-seven DS games to her friend, so she would have ninety - 
eighty seven = three left. The answer: three.`. Word2Num worked. Output: `3`

text2digits:
         - source:    Isabel gave eighty-seven DS games to her friend, so she would have ninety - eighty seven = 
one game left. The answer: one.
         - converted: Isabel gave 87 DS games to her friend, so she would have 90 - 87 = 1 game left. The answer: 1
         - final:     1

split_fn: no numbers found. Received: `Isabel gave eighty-seven DS games to her friend, so she would have ninety - 
eighty seven = one game left. The answer: one.`. Word2Num worked. Output: `1`

text2digits:
         - source:    Isabel gave eighty-seven games to her friend, so she would have ninety - eighty seven = three
left. The answer: three.
         - converted: Isabel gave 87 games to her friend, so she would have 90 - 87 = 3 left. The answer: 3
         - final:     3

split_fn: no numbers found. Received: `Isabel gave eighty-seven games to her friend, so she would have ninety - 
eighty seven = three left. The answer: three.`. Word2Num worked. Output: `3`

Accuracy: 37.7%:  45%|████▌     | 920/2023 [28:14<43:28,  2.37s/it]

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would be left with ninety -
seven = two cakes. The answer: two.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would be left with 90 - 7 = 2 cakes. The 
answer: 2
         - final:     2

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would be 
left with ninety - seven = two cakes. The answer: two.`. Word2Num worked. Output: `2`

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would be left with ninety -
seven = four cakes. The answer: four.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would be left with 90 - 7 = 4 cakes. The 
answer: 4
         - final:     4

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would be 
left with ninety - seven = four cakes. The answer: four.`. Word2Num worked. Output: `4`

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would still have ninety - 
seven = two cakes. The answer: two.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would still have 90 - 7 = 2 cakes. The 
answer: 2
         - final:     2

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would 
still have ninety - seven = two cakes. The answer: two.`. Word2Num worked. Output: `2`

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would be left with ninety -
seven = two cakes. The answer: two.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would be left with 90 - 7 = 2 cakes. The 
answer: 2
         - final:     2

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would be 
left with ninety - seven = two cakes. The answer: two.`. Word2Num worked. Output: `2`

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would be left with ninety -
seven = four cakes. The answer: four.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would be left with 90 - 7 = 4 cakes. The 
answer: 4
         - final:     4

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would be 
left with ninety - seven = four cakes. The answer: four.`. Word2Num worked. Output: `4`

text2digits:
         - source:    The baker sold ninety - eighty three = seven cakes. So, the baker would still have ninety - 
seven = two cakes. The answer: two.
         - converted: The baker sold 90 - 83 = 7 cakes. So, the baker would still have 90 - 7 = 2 cakes. The 
answer: 2
         - final:     2

split_fn: no numbers found. Received: `The baker sold ninety - eighty three = seven cakes. So, the baker would 
still have ninety - seven = two cakes. The answer: two.`. Word2Num worked. Output: `2`

Accuracy: 37.8%:  48%|████▊     | 965/2023 [29:35<27:57,  1.59s/it]

text2digits:
         - source:    The clown now has forty-seven + thirteen = sixty balloons. The answer: sixty.
         - converted: The clown now has 47 + 13 = 60 balloons. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The clown now has forty-seven + thirteen = sixty balloons. The answer: 
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown has forty-seven + thirteen = sixty balloons now. The answer: sixty.
         - converted: The clown has 47 + 13 = 60 balloons now. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The clown has forty-seven + thirteen = sixty balloons now. The answer: 
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown has forty-seven + thirteen = sixty two balloons now. The answer: sixty two.
         - converted: The clown has 47 + 13 = 62 balloons now. The answer: 62
         - final:     62

split_fn: no numbers found. Received: `The clown has forty-seven + thirteen = sixty two balloons now. The answer: 
sixty two.`. Word2Num worked. Output: `62`

text2digits:
         - source:    The clown blew up forty-seven + thirteen = sixty balloons now. So the answer is sixty.
         - converted: The clown blew up 47 + 13 = 60 balloons now. So the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The clown blew up forty-seven + thirteen = sixty balloons now. So the answer
is sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown blew up forty-seven + thirteen = sixty balloons. So, the answer is sixty.
         - converted: The clown blew up 47 + 13 = 60 balloons. So, the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The clown blew up forty-seven + thirteen = sixty balloons. So, the answer is
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown now has forty-seven + thirteen = sixty balloons. The answer: sixty.
         - converted: The clown now has 47 + 13 = 60 balloons. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The clown now has forty-seven + thirteen = sixty balloons. The answer: 
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown has forty-seven + thirteen = sixty balloons now. The answer: sixty.
         - converted: The clown has 47 + 13 = 60 balloons now. The answer: 60
         - final:     60

split_fn: no numbers found. Received: `The clown has forty-seven + thirteen = sixty balloons now. The answer: 
sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown has forty-seven + thirteen = sixty two balloons now. The answer: sixty two.
         - converted: The clown has 47 + 13 = 62 balloons now. The answer: 62
         - final:     62

split_fn: no numbers found. Received: `The clown has forty-seven + thirteen = sixty two balloons now. The answer: 
sixty two.`. Word2Num worked. Output: `62`

text2digits:
         - source:    The clown blew up forty-seven + thirteen = sixty balloons now. So the answer is sixty.
         - converted: The clown blew up 47 + 13 = 60 balloons now. So the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The clown blew up forty-seven + thirteen = sixty balloons now. So the answer
is sixty.`. Word2Num worked. Output: `60`

text2digits:
         - source:    The clown blew up forty-seven + thirteen = sixty balloons. So, the answer is sixty.
         - converted: The clown blew up 47 + 13 = 60 balloons. So, the answer is 60
         - final:     60

split_fn: no numbers found. Received: `The clown blew up forty-seven + thirteen = sixty balloons. So, the answer is
sixty.`. Word2Num worked. Output: `60`

Accuracy: 38.3%:  75%|███████▍  | 1512/2023 [46:26<14:37,  1.72s/it]

text2digits:
         - source:    If the farmer had ninety-seven tomatoes and picked eighty - three of them, how many did he 
have left? The answer: seven.
         - converted: If the farmer had 97 tomatoes and picked 80 - 3 of them, how many did he have left? The 
answer: 7
         - final:     7

split_fn: no numbers found. Received: `If the farmer had ninety-seven tomatoes and picked eighty - three of them, 
how many did he have left? The answer: seven.`. Word2Num worked. Output: `7`

text2digits:
         - source:    If the farmer had ninety-seven tomatoes and picked eighty - three of them, how many would be 
left? The answer: seven.
         - converted: If the farmer had 97 tomatoes and picked 80 - 3 of them, how many would be left? The answer: 
7
         - final:     7

split_fn: no numbers found. Received: `If the farmer had ninety-seven tomatoes and picked eighty - three of them, 
how many would be left? The answer: seven.`. Word2Num worked. Output: `7`

text2digits:
         - source:    If the farmer had ninety-seven tomatoes and picked eighty - three of them, how many did he 
have left? The answer: seven.
         - converted: If the farmer had 97 tomatoes and picked 80 - 3 of them, how many did he have left? The 
answer: 7
         - final:     7

split_fn: no numbers found. Received: `If the farmer had ninety-seven tomatoes and picked eighty - three of them, 
how many did he have left? The answer: seven.`. Word2Num worked. Output: `7`

text2digits:
         - source:    If the farmer had ninety-seven tomatoes and picked eighty - three of them, how many would be 
left? The answer: seven.
         - converted: If the farmer had 97 tomatoes and picked 80 - 3 of them, how many would be left? The answer: 
7
         - final:     7

split_fn: no numbers found. Received: `If the farmer had ninety-seven tomatoes and picked eighty - three of them, 
how many would be left? The answer: seven.`. Word2Num worked. Output: `7`

Accuracy: 38.4%:  76%|███████▌  | 1533/2023 [47:00<12:23,  1.52s/it]

text2digits:
         - source:    Our class got fifty - four + twenty three = eighty books from the library. The answer: 
eighty.
         - converted: Our class got 50 - 4 + 23 = 80 books from the library. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty - four + twenty three = eighty books from the library. 
The answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    Our class got fifty-four + twenty-three = eighty books from the library. The answer: eighty.
         - converted: Our class got 54 + 23 = 80 books from the library. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty-four + twenty-three = eighty books from the library. The
answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    Our class got fifty-four + twenty-three = eighty books from the library. So, the answer is 
eighty.
         - converted: Our class got 54 + 23 = 80 books from the library. So, the answer is 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty-four + twenty-three = eighty books from the library. So,
the answer is eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    Our class got fifty - four + twenty three = eighty books from the library. The answer: 
eighty.
         - converted: Our class got 50 - 4 + 23 = 80 books from the library. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty - four + twenty three = eighty books from the library. 
The answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    Our class got fifty-four + twenty-three = eighty books from the library. The answer: eighty.
         - converted: Our class got 54 + 23 = 80 books from the library. The answer: 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty-four + twenty-three = eighty books from the library. The
answer: eighty.`. Word2Num worked. Output: `80`

text2digits:
         - source:    Our class got fifty-four + twenty-three = eighty books from the library. So, the answer is 
eighty.
         - converted: Our class got 54 + 23 = 80 books from the library. So, the answer is 80
         - final:     80

split_fn: no numbers found. Received: `Our class got fifty-four + twenty-three = eighty books from the library. So,
the answer is eighty.`. Word2Num worked. Output: `80`

Accuracy: 38.7%:  82%|████████▏ | 1651/2023 [50:25<08:17,  1.34s/it]

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety two pieces of candy. The answer: ninety two.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 92
pieces of candy. The answer: 92
         - final:     92

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety two pieces of candy. The answer: ninety two.`. Word2Num worked. Output: 
`92`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety two pieces. The answer: ninety two.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 92
pieces. The answer: 92
         - final:     92

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety two pieces. The answer: ninety two.`. Word2Num worked. Output: `92`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety eight pieces of candy. The answer: ninety eight.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 98
pieces of candy. The answer: 98
         - final:     98

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety eight pieces of candy. The answer: ninety eight.`. Word2Num worked. 
Output: `98`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety eight pieces. The answer: ninety eight.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 98
pieces. The answer: 98
         - final:     98

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety eight pieces. The answer: ninety eight.`. Word2Num worked. Output: `98`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety two pieces of candy. The answer: ninety two.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 92
pieces of candy. The answer: 92
         - final:     92

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety two pieces of candy. The answer: ninety two.`. Word2Num worked. Output: 
`92`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety two pieces. The answer: ninety two.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 92
pieces. The answer: 92
         - final:     92

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety two pieces. The answer: ninety two.`. Word2Num worked. Output: `92`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety eight pieces of candy. The answer: ninety eight.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 98
pieces of candy. The answer: 98
         - final:     98

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety eight pieces of candy. The answer: ninety eight.`. Word2Num worked. 
Output: `98`

text2digits:
         - source:    Isabel had sixty - one pieces left after giving her friend candy, so she started with sixty +
twenty seven = ninety eight pieces. The answer: ninety eight.
         - converted: Isabel had 60 - 1 pieces left after giving her friend candy, so she started with 60 + 27 = 98
pieces. The answer: 98
         - final:     98

split_fn: no numbers found. Received: `Isabel had sixty - one pieces left after giving her friend candy, so she 
started with sixty + twenty seven = ninety eight pieces. The answer: ninety eight.`. Word2Num worked. Output: `98`

Accuracy: 38.0%:  91%|█████████ | 1840/2023 [56:22<04:35,  1.51s/it]

text2digits:
         - source:    Ellen has nine + six = fifteen balls. The answer: fifteen.
         - converted: Ellen has 9 + 6 = 15 balls. The answer: 15
         - final:     15

split_fn: no numbers found. Received: `Ellen has nine + six = fifteen balls. The answer: fifteen.`. Word2Num 
worked. Output: `15`

text2digits:
         - source:    Ellen has nine + six = fifteen balls. The answer: fifteen.
         - converted: Ellen has 9 + 6 = 15 balls. The answer: 15
         - final:     15

split_fn: no numbers found. Received: `Ellen has nine + six = fifteen balls. The answer: fifteen.`. Word2Num 
worked. Output: `15`

Accuracy: 38.1%:  91%|█████████▏| 1849/2023 [56:39<04:58,  1.71s/it]

text2digits:
         - source:    The difference between the number of hotdogs before selling and after selling is ninety - 
twenty = seventy. So, the answer is seventy.
         - converted: The difference between the number of hotdogs before selling and after selling is 90 - 20 = 
70. So, the answer is 70
         - final:     70

split_fn: no numbers found. Received: `The difference between the number of hotdogs before selling and after 
selling is ninety - twenty = seventy. So, the answer is seventy.`. Word2Num worked. Output: `70`

text2digits:
         - source:    The difference between the number of hotdogs before and after selling is ninety - twenty = 
seventy. So, the answer is seventy.
         - converted: The difference between the number of hotdogs before and after selling is 90 - 20 = 70. So, 
the answer is 70
         - final:     70

split_fn: no numbers found. Received: `The difference between the number of hotdogs before and after selling is 
ninety - twenty = seventy. So, the answer is seventy.`. Word2Num worked. Output: `70`

text2digits:
         - source:    The difference between the number of hotdogs before selling and after selling is ninety - 
twenty = seventy. So, the answer is seventy.
         - converted: The difference between the number of hotdogs before selling and after selling is 90 - 20 = 
70. So, the answer is 70
         - final:     70

split_fn: no numbers found. Received: `The difference between the number of hotdogs before selling and after 
selling is ninety - twenty = seventy. So, the answer is seventy.`. Word2Num worked. Output: `70`

text2digits:
         - source:    The difference between the number of hotdogs before and after selling is ninety - twenty = 
seventy. So, the answer is seventy.
         - converted: The difference between the number of hotdogs before and after selling is 90 - 20 = 70. So, 
the answer is 70
         - final:     70

split_fn: no numbers found. Received: `The difference between the number of hotdogs before and after selling is 
ninety - twenty = seventy. So, the answer is seventy.`. Word2Num worked. Output: `70`

Accuracy: 37.9%: 100%|██████████| 2023/2023 [1:02:24<00:00,  1.85s/it]


{
    'shuffle': True,
    'verbose': False,
    'num_beams': 8,
    'batch_size': 1,
    'max_new_tokens': 100,
    'use_majority_vote': True,
    'use_group_beam_search': False,
    'generation_extra_kwargs': {'repetition_penalty': 50.0},
    'context': '',
    'with_scratchpads': True
}

Accuracy, torch.float32: 37.9%

### ASDiv Dataset:

- **At a glance**: float32, no context, 8 beams majority vote, fixed number parser.
  - **Accuracy**: 37.9%
  - **Precision**: float32
  - **num_beams**: 8
  - **batch_size**: 1
  - **max_new_tokens**: 100
  - **use_majority_vote**: True
  - **use_group_beam_search**: False
  - **generation_extra_kwargs**: 
    - *repetition_penalty*: 50.0
    - *context*: ''

-> Uses "Let's do this step by step"

**Notes**:
    For the word outputs, the word2num often doesn't pickup the correct (last) word.
